src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs

```diff
@@ -8247,6 +8247,186 @@ static void Test<T>((T, T, T) x)
+                // (17,15): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T2' in the generic type or method 'ValueTuple<T1, T2>'
+                //         (int, int) t1 = t0
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "int").WithArguments("System.ValueTuple<T1, T2>", "T2", "int").WithLocation(17, 15)
+            );
+        }
+
+        [Fact]
+        [WorkItem(15399, "https://github.com/dotnet/roslyn/issues/15399")]
+        public void Constraints_02()
+        {
+            var source = @"
+using System;
+class Program
+{
+    unsafe void M((int, int*) p, ValueTuple<int, int*> q)
+    {
+        (int, int*) t0 = p;
+        var t1 = (1, (int*)null);
+        var t2 = new ValueTuple<int, int*>(1, null);
+        ValueTuple<int, int*> t3 = t2;
+    }
+}";
+            var comp = CreateCompilationWithMscorlib45AndCSruntime(source, additionalRefs: s_valueTupleRefs,
+               options: TestOptions.UnsafeDebugDll);
+            comp.VerifyDiagnostics(
+                // (5,31): error CS0306: The type 'int*' may not be used as a type argument
+                //     unsafe void M((int, int*) p, ValueTuple<int, int*> q)
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "p").WithArguments("int*").WithLocation(5, 31),
+                // (5,56): error CS0306: The type 'int*' may not be used as a type argument
+                //     unsafe void M((int, int*) p, ValueTuple<int, int*> q)
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "q").WithArguments("int*").WithLocation(5, 56),
+                // (7,15): error CS0306: The type 'int*' may not be used as a type argument
+                //         (int, int*) t0 = p;
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "int*").WithArguments("int*").WithLocation(7, 15),
+                // (8,22): error CS0306: The type 'int*' may not be used as a type argument
+                //         var t1 = (1, (int*)null);
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "(int*)null").WithArguments("int*").WithLocation(8, 22),
+                // (9,38): error CS0306: The type 'int*' may not be used as a type argument
+                //         var t2 = new ValueTuple<int, int*>(1, null);
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "int*").WithArguments("int*").WithLocation(9, 38),
+                // (10,25): error CS0306: The type 'int*' may not be used as a type argument
+                //         ValueTuple<int, int*> t3 = t2;
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "int*").WithArguments("int*").WithLocation(10, 25)
+            );
+        }
+
+        [Fact]
+        public void Constraints_03()
+        {
+            var source = @"
+using System.Collections.Generic;
+namespace System
+{
+    public struct ValueTuple<T1, T2>
+        where T2 : class
+    {
+        public ValueTuple(T1 _1, T2 _2)
+        {
+        }
+    }
+}
+class C<T>
```

Comment:
I think he means: `class C<T> where T : class ...` and `(U, U) M<U>(U x) where U : class ...`.

---
In reply to: [92733403](https://github.com/dotnet/roslyn/pull/15641#discussion_r92733403) [](ancestors = 92733403)